packages/blockchain-tools/src/contracts/soroban/credits/src/contract.rs

```diff
@@ -100,7 +100,7 @@ impl Credits {
     events::donation(&e, from, provider, amount);
   }
 
-  pub fn swap_xlm_to_carbon(e: Env, from: Address, xlm_amount: i128) {
+  pub fn swap_xlm_to_carbon(e: Env, from: Address, xlm_amount: i128) -> i128 {
     if xlm_amount <= 0 { panic!("amount less than zero") }
     from.require_auth();
 
```

Comment:
I'd move all this inner logic to a new `sink.rs` module, except for the `from.require_auth()` call. When this contract does the swap, it should not ask for additional auth. When there is an external caller, it does need auth for the `from` account.